rfc/001-composable-pipelines/README.md

```diff
@@ -0,0 +1,364 @@
+# Composable Pipelines
+
+- **Status**: In-Review
+- **Created**: Aug 19, 2022
+- **ETA**: Sep 21, 2022
+- **Authors**:
+  - [Matthias Vallentin](https://github.com/mavam)
+- **Contributors**:
+  - [Anthony Verez](https://github.com/netantho)
+  - [Dominik Lohmann](https://github.com/dominiklohmann)
+  - [Rémi Dettai](https://github.com/rdettai)
+  - [Thomas Peiselt](https://github.com/dispanser)
+  - [Tobias Mayer](https://github.com/tobim)
+- **Discussion**: [PR #2511](https://github.com/tenzir/vast/pull/2511)
+
+## Overview
+
+This proposal ideates a new UX around VAST's pipelines. The fundamental idea is
+that *everything is a pipeline* and build a composable UX around this concept.
+
+## Problem Statement
+
+Pipelines offer flexible data reshaping at [predefined customization
+points](https://vast.io/docs/use-vast/transform). For example, they can be
+triggered for specific events on import/export, or invoked by compaction.
+Howerver, it is impossible to run pipelines in an ad-hoc manner, for example as
+a post-processing step after a query or for an ad-hoc ingest of a data set.
+
+This RFC proposes dynamic pipelines that do not require static definition in
+configuration file, but rather allow users to specify them as an extension to
+expression language.
+
+## Solution Proposal
+
+Assume for a moment everything is a pipeline operator. Then we may write:
+
+ ```bash
+vast 'load s3://aws json |
+      summarize count(dst) group-by src |
+      store parquet /path/to/directory'
+```
+
+The UX would translate seamlessly to other languages, e.g., Python:
+
+```python
+import vast as v
+await v.load_json("s3://aws")
+       .group_by("src") # could also be merged with summarize()
+       .summarize(v.count("dst"))
+       .store_parquet("/path/to/directory")
+```
+
+The core of this proposal introduces pipeline operators as functions with input
+and output types. The following overview introduces the pipeline building
+blocks used throughout this proposal.
+
+A pipeline operator has an input and output type, e.g., `load<In, Out>` takes as
+input `In` and produces instances of `Out`. We consider the following valid
+types at the logical level:
+
+1. `Void`: no input or output, to represent a side effect
+2. `Arrow`: table slices (i.e., record batches)
```

Comment:
Now that we only have two types left at the logical level, it seems that the only thing that's really left is `Arrow`, as `Void` is only relevant for start and end of the pipeline, aka `source` and `sink`.  So the `<In, Out>` pairs make a lot of sense if there's more than one intermediate type, but with only one of those becomes unnecessary.

The alternative would be to only talk about `operators` (taking and producing `Arrow`), and `source` , `sink`, skipping the entire notion of `<In, Out>`. Just wanted to mention this, but I have no strong preferences here.

Reply:
Good observation. I added a note that we want to consider a simplified model at the logical level.